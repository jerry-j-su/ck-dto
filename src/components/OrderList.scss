@use 'sass:color';
@use '../theme.scss';

.order-list-container {
    &.hidden {
        // visibility: hidden!important;
        // height: 0!important;
        display:none;
    }

    height: 100%;
    overflow: hidden auto;
    scrollbar-gutter: stable both-edges; // very important, prevent visible scroll bar causing inner content to reflow/repaint
    &::-webkit-scrollbar {
        // display: none;
        width: 5px;
        height: 5px;
    }
    &::-webkit-scrollbar-track {
        border-radius: 2px;
        background-color: color.scale(theme.$bg-front, $whiteness: 7%);
    }
    &::-webkit-scrollbar-thumb {
        border-radius: 2px;
        background-color: theme.$font-primary-strong;
    }

    .order-count {
        font-size: .8rem;
        color: theme.$font-fade;
    }

    .order-list-body {
        padding: 0;
        margin-block: 0;
        list-style-type: none;
        display: grid;
        flex-direction: column;
    }

    .order-wrapper {
        border: 1px solid color.scale(theme.$border-fade, $blackness: 30%);
        border-radius: .2px;
        margin: .2rem;

        dl {
            display: flex;
            flex-wrap: wrap;
            margin: 0;
            padding: .5rem 0;

            .top-row,
            .col-1,
            .col-2 {
                padding: .2rem .5rem;
            }

            .top-row {
                flex: 1 0 100%;
            }

            .col-1 {
                flex: 0 1 15%;
                display: flex;
                justify-content: center;
                align-items: center;
            }

            .col-2 {
                flex: 1 0 auto;
            }

            .label {
                font-size: .6rem;
                color: theme.$font-fade;
                vertical-align: middle;
            }

            .value {
                &.name {
                    font-weight: 700;
                }

                &.status,
                &.items {
                    font-size: .8rem;
                }

                &.name {
                    color: theme.$font-primary-strong;
                }

                &.price {
                    color: theme.$highlight;
                }
            }

            dt {
                display: inline;
            }

            dd {
                display: inline;
                margin-inline-start: .5rem;
                margin-bottom: .2rem;
            }

            .name-spacer {
                flex: 1 0 90%;
                border-bottom: 1px dashed color.scale(theme.$border-fade, $blackness: 50%);
                margin: 0 1rem 0.2rem 1rem;
            }

            .status-spacer {
                flex: 0 0 auto;
                border-left: 1px dashed color.scale(theme.$border-fade, $blackness: 50%);
                margin: .7rem 0;
            }

            .spacer {
                margin-block: .5rem;
            }
        }
    }
}

.scrollable-wrapper {
    width: 100%;
    min-height: 0;
    position: relative;
    .scrollable-tip {
        &-top {
            position: absolute;
            top: 0;
        }
        &-bottom {
            position: absolute;
            bottom: 0;
        }
    }
}
