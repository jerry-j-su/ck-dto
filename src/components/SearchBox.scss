@use 'sass:color';
@use '../theme.scss';

.search-container {
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    .search-placeholder {
        position: absolute;
        margin-inline: .7rem;
        line-height: 2rem;
        cursor: default;
        pointer-events: none;
        color: color.scale(theme.$font-primary-strong, $blackness: 30%);
    }

    .search-input {
        flex: 1 0 100%;
        min-height: 2rem;
        display: flex;
        margin: 1rem 0rem .2rem;
        border: 1px solid #ccc;
        position: relative;
        &:focus-within {
            outline: -webkit-focus-ring-color auto 1px;
        }
        input {
            justify-self: flex-start;
            flex: 1;
            display: block;
            min-width: 30rem;
            padding: 0 .5rem;
            border: none;
            font-size: 1rem;
            &:focus-visible {
                outline: none;
            }
            color: theme.$font-primary-strong;
            background-color: color.scale(theme.$dark-bg, $whiteness: 10%);
        }
        .search-clear {
            line-height: 2rem;
            padding: 0 1rem;
            justify-self: flex-end;
            cursor :pointer;
            color: theme.$font-primary-strong;
            font-weight: 700;
            &:hover {
                color: theme.$secondary;
                background-color: theme.$font-primary-strong;
            }
        }
    }

    .search-notification {
        padding: 0 .6rem;
        color: color.scale(theme.$color-warn-primary, $whiteness: 30%);
    }
}
