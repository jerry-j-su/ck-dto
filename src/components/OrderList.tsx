import React, { useState, useEffect, useMemo, forwardRef } from 'react'

import { OrderType } from '../types'
import { ScrollableContainerExtraProps } from './dynamicScrollableListHOC'
import OrderSkeleton from './OrderSkeleton'
import { mergeClassNames } from '../utils'
import './OrderList.scss'

type OrderListProps = {
    orders: OrderType[],
    containerClassName?: string,
}
const UpdateBatchSize = 10  // how many buffered records to attach to DOM tree if user scroll down to the bottom of the page

export default forwardRef(({ orders, containerClassName, holdAppendChild }: OrderListProps & ScrollableContainerExtraProps, ref: any) => {
    const [renderCount, setRenderCount] = useState<number>(0)
    useEffect(() => {
        if (holdAppendChild) return
        if (renderCount < orders.length) {
            setRenderCount(renderCount + UpdateBatchSize)
        }
    }, [holdAppendChild, orders, orders.length, renderCount, setRenderCount])

    const ordersToShow = useMemo(() => {
        return orders.slice(0, renderCount)
    }, [orders, orders.length, renderCount]) // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <section className={mergeClassNames(['order-list-container', containerClassName])} ref={ref} tabIndex={0}>
            <ul className="order-list-body">
                {ordersToShow.map(({ id, customer, destination, 'event_name': status, item, price, 'sent_at_second': time }) => (
                    <li className="order-wrapper" key={`${id}-${status}`}>
                            <div className="top-row">
                                <span className="label name">name</span>
                                <span className="value name">{customer}</span>
                            </div>
                            <div className="name-spacer" />
                            <div className="col-1">
                                <span className="label status" style={{ display: 'none' }}>Status</span>
                                <span className="value status">{status}</span>
                            </div>
                            <div className="status-spacer" />
                            <div className="col-2">
                                <span className="label items">items</span>
                                <span className="value items">{item}</span>
                                <div className="spacer" />
                                <span className="label price">price</span>
                                <span className="value price">{price}</span>
                            </div>
                    </li>
                ))}
                {/* DOM append is held due, display skeleton */}
                {holdAppendChild && renderCount < orders.length
                    ? (<>
                        <li className="order-wrapper skeleton"><OrderSkeleton /></li>
                        <li className="order-wrapper skeleton"><OrderSkeleton /></li>
                      </>)
                    : <li className="order-wrapper end"><div className="loading-dot" /></li>
                }
            </ul>
        </section>
    )
})
