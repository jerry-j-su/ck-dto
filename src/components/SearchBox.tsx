import React, {
useCallback, useState, useRef,
FormEvent, FormEventHandler,
} from 'react'

import { tailingDebounce } from '../utils'
import useOrders from '../hooks/useOrderFlow'
import './SearchBox.scss'

/**
 * Validate price input value
 * Currently it only allows pure number
 * @param {string} rawValue
 */
function validatePriceInput(rawValue: string) {
    return /^\s*\d*\.?\d*\s*$/.test(rawValue)
}

export function InputBox({ value, onInput, clearInput }: { value: string, onInput: FormEventHandler, clearInput: FormEventHandler }) {
    const inputDOM = useRef<HTMLInputElement>(null)
    return (
        <div className="search-input" tabIndex={0}>
            {!(inputDOM.current && inputDOM.current.value) && <div className="search-placeholder">Enter price to search among orders...</div>}
            <input type="combobox" value={value} onInput={onInput} ref={inputDOM} />
            {value && <div className="search-clear" onClick={clearInput}>X</div>}
        </div>
    )
}

export default function SearchBox() {
    const [searchStr, setSearchStr] = useState<string>('')
    const [warnMsg, setWarnMsg] = useState<string>('')
    const { setFilterCriteria } = useOrders()

    /**
     * Perform an order search, debounced, tailing, 500ms
     */
    const performSearch = useCallback( // eslint-disable-line react-hooks/exhaustive-deps
        tailingDebounce((price?: number) => {
            console.log(`searching price: ${price}`)
            setFilterCriteria({ price })
        }, 500),
        [setFilterCriteria]
    )

    /* Handlers */
    const clear = useCallback(() => {
        setSearchStr('')
        setWarnMsg('')
        performSearch()
    }, [setSearchStr, performSearch])

    /**
     * Handler to parse and validate search input
     */
    const onSearchInput: FormEventHandler = useCallback((event: FormEvent) => {
        // @ts-ignore
        const { target: { value: rawInput } } = event
        setSearchStr(rawInput)
        if (validatePriceInput(rawInput)){
            setWarnMsg('')
            const price = parseInt(rawInput) || undefined
            performSearch(price)
        } else {
            setWarnMsg('Invalid price amount')
        }
    }, [setSearchStr, performSearch])

    return (
        <section className="search-container">
            <InputBox value={searchStr} onInput={onSearchInput} clearInput={clear} />
            {warnMsg && <div className="search-notification">{warnMsg}</div>}
        </section>
    )
}
