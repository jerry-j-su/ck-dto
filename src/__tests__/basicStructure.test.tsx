import React from 'react'
import { render } from '@testing-library/react'
import { readFileSync  } from 'fs';

import SearchBox, { InputBox } from '../components/SearchBox'
import OrderList from '../components/OrderList'


// module mocks
jest.mock('react', () => {
    const originalModule = jest.requireActual('react')
    return {
        __esModule: true,
        ...originalModule,
        useRef: () => ({ current: { value: 'random content' } }),
        useState: (initialValue: any) => [initialValue, () => {}],
        useCallback: () => () => {},
        useEffect: () => {},
    }
})
jest.mock('../state/useOrders', () => ({
    __esModule: true,
    default: () => ({
        orderList: mockOrders,
        setFilterCriteria: () => { console.log('set filter criteria ')},
    })
}))
// data mocks
const mockOrders = JSON.parse(readFileSync(`${__dirname}/dummyOrders.json`, { encoding: 'utf-8' }))

describe('High-level Component render before actual DOM', () => {
    test('SearchBox Input box basic render', () => {
        const onInput = jest.fn()
        const clearInput = jest.fn()
        const nodes = render(InputBox({ value: '3', onInput, clearInput }))
        expect(nodes.container).toMatchSnapshot()
    })

    test('SearchBox basic render', () => {
        const nodes = render(SearchBox())
        expect(nodes.container).toMatchSnapshot()
    })

    test('Order List basic render', () => {
        const nodeTree = render(OrderList())
        expect(nodeTree.container).toMatchSnapshot()
    })
})
