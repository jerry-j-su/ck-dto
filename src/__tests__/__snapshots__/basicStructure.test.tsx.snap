// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`High-level Component render before actual DOM SearchBox Input box basic render 1`] = `
<div>
  <div
    class="search-input"
  >
    <input
      aria-errormessage="search-error-msg"
      aria-invalid="false"
      aria-label="Input price to search among orders"
      type="text"
      value="3"
    />
    <button
      aria-label="clear search"
      class="search-clear"
      type="button"
    >
      X
    </button>
  </div>
</div>
`;

exports[`High-level Component render before actual DOM SearchBox basic render 1`] = `
<div>
  <section
    class="search-container"
  >
    <div
      class="search-input"
    >
      <input
        aria-errormessage="search-error-msg"
        aria-invalid="false"
        aria-label="Input price to search among orders"
        type="text"
        value=""
      />
    </div>
  </section>
</div>
`;
